packages/flutter/lib/src/foundation/assertions.dart

```diff
@@ -3,19 +3,114 @@
+         showName: false,
+         showSeparator: false,
+         defaultValue: null,
+         style: style,
+         level: level,
+       );
+
+  String valueToString({ TextTreeConfiguration parentConfiguration }) {
+    return value.join('');
+  }
+}
+
+/// Diagnostics describing an error message.
+///
+/// An error message should start with a summary that is typically
+/// no longer than 2 lines that concisely states what the assertion
+/// failure or contract violation was. The summary should be separated from the
+/// rest of the error message with a line break.
+///
+/// The client (e.g., an IDE) usually displays the error summary in red.
+///
+/// The full error message should include at least the following elements:
+/// * Claim: Explanation of the assertion failure or contract violation.
+/// * Grounds: Facts about the user's code that led to the error.
+/// * Warranty: Connections between the grounds and the claim.
+///
+/// Error messages may include references to objects relevant to the error using
+/// string interpolation and those objects will be directly available to
+/// debugging tools when error messages are generating as part of a debug builds.
+/// Other builds will see the same string error message but will not be able to
+/// extract the contents of the object.
+class ErrorMessage extends _ErrorDiagnostic {
+  // This constructor provides a reliable hook for a kernel transformer to find
+  // error messages that need to be rewritten to include object references.
+  //
+  // The message will display with the same text regardless of whether the
+  // kernel transformer is used but without the kernel transformer, debugging
+  // tools are unable to provide interactive displays of objects referenced by
+  // the string.
+  /// In debug builds, a kernel transformer rewrites calls to the default
+  /// constructor into calls to this constructor.
+  ///
+  /// of this class as follows so that the objects referenced in the
+  /// string literal can be captured by debuggers.
+  /// ```dart
+  /// ErrorMessage('Element $element must be color $color')
+  /// ```
+  /// Desugars to:
+  /// ```dart
+  /// ErrorMessage._structured(<Object>['Element ', element, ' must be ', color]);
+  /// ```
+  ///
+  /// Slightly more complext case:
+  /// ```dart
+  /// ErrorMessage('Element ${element.runtimeType} must be must be $color')
+  /// ```
+  /// Desugars to:
+  ///```dart
+  /// ErrorMessage._structured(<Object>[
+  ///   'Element ',
+  ///    DiagnosticsProperty(null, element, description: element.runtimeType?.toString()),
+  ///    ' must be ',
+  ///    color,
+  /// ]);
+  /// ```
+  ErrorMessage(String message) : super._fromParts([message], level: DiagnosticLevel.summary);
+
+  // Constructor that calls to the default constructor are rewriten to use in
+  // debug mode using a kernel transformer.
+  ErrorMessage._fromParts(List<Object> messageParts) : super._fromParts(messageParts, level: DiagnosticLevel.summary);
```

Comment:
Is this supposed to be `ErrorMessage._structured` maybe?

Reply:
I went back and forth on the name. Any preference on the name for the private constructor used by the kernel transformer?

Reply:
Slightly leaning to `_fromParts` - but not really a strong preference since users will presumably never see this?